crates/storage/src/set.rs

```diff
@@ -150,3 +148,397 @@ where
+                    Order::Ascending,
+                )
+                .collect::<StdResult<Vec<_>>>()?;
+
+            assert_eq!(data.len(), 30);
+            assert_eq!(data[0], concat(b"foo", &0_u32.joined_prefix()));
+            assert_eq!(data[29], concat(b"foo", &29_u32.joined_prefix()));
+        }
+
+        // Max Min bound
+        {
+            let data = SINGLE
+                .range(
+                    storage,
+                    Some(Bound::inclusive(
+                        concat(b"foo", &40_u32.joined_prefix()).as_slice(),
+                    )),
+                    Some(Bound::exclusive(
+                        concat(b"foo", &50_u32.joined_prefix()).as_slice(),
+                    )),
+                    Order::Ascending,
+                )
+                .collect::<StdResult<Vec<_>>>()?;
+
+            assert_eq!(data.len(), 10);
+            assert_eq!(data[0], concat(b"foo", &40_u32.joined_prefix()));
+            assert_eq!(data[9], concat(b"foo", &49_u32.joined_prefix()));
+        }
+
+        Ok(())
+    }
+
+    #[test]
+    fn decimal_range() -> StdResult<()> {
+        let storage = &mut MockStorage::new();
+
+        for i in -50..50 {
+            DOUBLE.insert(storage, (Dec128::from_str(&i.to_string())?, &i.to_string()))?;
+        }
+
+        let data = DOUBLE.all(storage)?;
+
+        assert_eq!(data.len(), 100);
+
+        for (index, val) in (-50..50).enumerate() {
+            assert_eq!(
+                data[index],
+                (Dec128::from_str(&val.to_string()).unwrap(), val.to_string())
+            );
+        }
+
+        Ok(())
+    }
+
+    #[test]
+    fn prefix() -> StdResult<()> {
+        let storage = &mut MockStorage::new();
+
+        DOUBLE.insert(storage, (Dec128::from_str("-2")?, "a"))?;
+        DOUBLE.insert(storage, (Dec128::from_str("-2")?, "b"))?;
+        DOUBLE.insert(storage, (Dec128::from_str("-2")?, "c"))?;
+        DOUBLE.insert(storage, (Dec128::from_str("-2")?, "d"))?;
+        DOUBLE.insert(storage, (Dec128::from_str("-2")?, "e"))?;
+        DOUBLE.insert(storage, (Dec128::from_str("-1.5")?, "a"))?;
+        DOUBLE.insert(storage, (Dec128::from_str("-1.5")?, "b"))?;
+        DOUBLE.insert(storage, (Dec128::from_str("0")?, "abcb"))?;
+        DOUBLE.insert(storage, (Dec128::from_str("1")?, "b"))?;
+        DOUBLE.insert(storage, (Dec128::from_str("2")?, "b"))?;
+        DOUBLE.insert(storage, (Dec128::from_str("2")?, "a"))?;
+
+        // No bound
+        {
+            let val = DOUBLE
+                .prefix(Dec128::from_str("-2")?)
+                .keys(storage, None, None, Order::Ascending)
+                .collect::<StdResult<Vec<_>>>()?;
+
+            assert_eq!(val.len(), 5);
+            assert_eq!(val[0], "a".to_string());
+            assert_eq!(val[1], "b".to_string());
+            assert_eq!(val[2], "c".to_string());
+            assert_eq!(val[3], "d".to_string());
+            assert_eq!(val[4], "e".to_string());
```

Comment:
This can be simplified as just

```rust
assert_eq!(val, ["a", "b", "c", "d", "e"]);
```

Apparently, `Vec<String>` implements `PartialEq<[&str; N]>` which is handy.